[TensorBoard] Lazy initialize summary writers everywhere

kunalb opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100319

I've seen a lot of code that eagerly initializes a SummaryWriter instance with a logdir: which immediately creates a 40 byte file with a timestamp and file version, and then never uses it.

This is prevalent to the point that I could speed up our instance of TB significantly by simply skipping 40 byte files that had events in the name (some frameworks create them at 1 per rank); it seems worth adjusting this behavior globally for all PyTorch users.

Differential Revision: [D45425098](https://our.internmc.facebook.com/intern/diff/D45425098/)

Commit:
[TensorBoard] Lazy initialize summary writers everywhere

I've seen a lot of code that eagerly initializes a SummaryWriter instance with a logdir: which immediately creates a 40 byte file with a timestamp and file version, and then never uses it.

This is prevalent to the point that I could speed up our instance of TB significantly by simply skipping 40 byte files that had events in the name (some frameworks create them at 1 per rank); it seems worth adjusting this behavior globally for all PyTorch users.

Differential Revision: [D45425098](https://our.internmc.facebook.com/intern/diff/D45425098/)

[ghstack-poisoned]

Commit:
Update on "[TensorBoard] Lazy initialize summary writers everywhere"


I've seen a lot of code that eagerly initializes a SummaryWriter instance with a logdir: which immediately creates a 40 byte file with a timestamp and file version, and then never uses it.

This is prevalent to the point that I could speed up our instance of TB significantly by simply skipping 40 byte files that had events in the name (some frameworks create them at 1 per rank); it seems worth adjusting this behavior globally for all PyTorch users.

Differential Revision: [D45425098](https://our.internmc.facebook.com/intern/diff/D45425098/)

[ghstack-poisoned]

kunalb commented:
I think I can simplify this a little bit by keeping SummaryWriter lazy (it's almost there) and leaving FileWriter as is. Updating.

Skylion007 commented on torch/utils/tensorboard/writer.py:
```diff
@@ -73,13 +74,24 @@ def __init__(self, log_dir, max_queue=10, flush_secs=120, filename_suffix=""):
+        self._max_queue = max_queue
+        self._flush_secs = flush_secs
+        self._filename_suffix = filename_suffix
+        self._event_writer = None
```
```suggestion
        self._event_writer : Optional[EventFileWriter] = None
```
So it's not an implicit optional

Commit:
Update on "[TensorBoard] Lazy initialize summary writers everywhere"


I've seen a lot of code that eagerly initializes a SummaryWriter instance with a logdir: which immediately creates a 40 byte file with a timestamp and file version, and then never uses it.

This is prevalent to the point that I could speed up our instance of TB significantly by simply skipping 40 byte files that had events in the name (some frameworks create them at 1 per rank); it seems worth adjusting this behavior globally for all PyTorch users.

Differential Revision: [D45425098](https://our.internmc.facebook.com/intern/diff/D45425098/)

[ghstack-poisoned]

Commit:
Update on "[TensorBoard] Lazy initialize summary writers everywhere"


I've seen a lot of code that eagerly initializes a SummaryWriter instance with a logdir: which immediately creates a 40 byte file with a timestamp and file version, and then never uses it.

This is prevalent to the point that I could speed up our instance of TB significantly by simply skipping 40 byte files that had events in the name (some frameworks create them at 1 per rank); it seems worth adjusting this behavior globally for all PyTorch users.

Differential Revision: [D45425098](https://our.internmc.facebook.com/intern/diff/D45425098/)

[ghstack-poisoned]

kunalb commented:
Hm, given the test failure I wonder if I should leave .file_writer introspectable to maintain the current behavior of checking .file_writer is None to see if the SummaryWriter was closed.

kunalb commented:
(will fix up in a bit, forgot to follow up here)
